sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala

```diff
@@ -219,10 +222,35 @@ object StreamingSymmetricHashJoinHelper extends Logging {
           attributesWithEventWatermark = AttributeSet(otherSideInputAttributes),
           condition,
           eventTimeWatermarkForEviction)
-        val inputAttributeWithWatermark = oneSideInputAttributes.find(_.metadata.contains(delayKey))
-        val expr = watermarkExpression(inputAttributeWithWatermark, stateValueWatermark)
-        expr.map(JoinStateValueWatermarkPredicate.apply _)
 
+        // For example, if the condition is of the form:
+        //    left_time > right_time + INTERVAL 30 MINUTES
+        // Then this extracts left_time and right_time.
+        val attributesInCondition = AttributeSet(
+          condition.get.collect { case a: AttributeReference => a }
+        )
+
+        // Construct an AttributeSet so that we can perform equality between attributes,
+        // which we do in the filter below.
+        val oneSideInputAttributeSet = AttributeSet(oneSideInputAttributes)
+
+        // oneSideInputAttributes could be [left_value, left_time], and we just
+        // want the attribute _in_ the time-interval condition.
+        val oneSideStateWatermarkAttributes = attributesInCondition.filter { a =>
+            oneSideInputAttributeSet.contains(a)
+        }
+
+        // There should be a single attribute per side in the time-interval condition, so,
```

Comment:
Where is this part ensured? ('single attribute per side').

Reply:
I'm having a hard time finding exactly where we specify the grammar for the join `Expression`, but empirically, the grammar is something like: `ATTRIBUTE`, `COMPARATOR`, `ATTRIBUTE`, and then an `INTERVAL`. So, you can't specify more than 2 attributes without getting an exception.

We also make this assumption in higher up in `StreamingJoinHelper.getStateValueWatermark` as well: https://github.com/apache/spark/blob/f3a6ca9e2c476fe418679f82b0c5148d69d0fbda/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala#L142